# Brussels Airlines Special Flights during Disruption

In [76]:
from json import loads, dumps
import pandas as pd
import numpy as np
# EXTRACT
with open("all-flights.json") as f1:
    sched_flights = pd.DataFrame(loads(f1.read())["AirlineFlightSchedulesResult"]["data"])

with open("historical_flights.json") as f2:
    hist_flights = pd.DataFrame(loads(f2.read())).T

with open("airports-new.json") as f3:
    airports = loads(f3.read())
    ap = pd.DataFrame(airports).T

airports is a python dict, which is easier for cycling through all airports and looking up individual airports by ICAO code. The Pandas dataframe is easier for joining it with other dataframes.

In [77]:
# TRANSFORM
import arrow

sched_flights['arrowdate'] = sched_flights.departuretime.apply(arrow.get)
sched_flights['date'] = sched_flights.arrowdate.apply(lambda x: x.format('YYYY-MM-DD'))
bel_sched_flights = sched_flights[sched_flights["actual_ident"]==""]
bel_sched_flights = bel_sched_flights[["ident","aircrafttype", "origin", "destination", \
                           "seats_cabin_coach", "date", "departuretime"]]

We'll perform a transformation to determine the city pairs like this. Similarly, we save the seat capacity per aircraft type.

In [78]:
citypairs = bel_sched_flights[["origin", "destination"]].drop_duplicates()

In [79]:
ac = bel_sched_flights.groupby(["aircrafttype"])
aircraft = ac.aggregate(np.mean)['seats_cabin_coach']

Let's calculate the distances for the city pairs

In [80]:
import numpy as np
from numpy import pi
def distance(lat1, lon1, lat2, lon2):
    # http://www.movable-type.co.uk/scripts/latlong.html
    R = 6371 / 1.6 # in status miles
    lat1 *= pi/180.
    lon1 *= pi/180.
    lat2 *= pi/180.
    lon2 *= pi/180.
    return R*np.arccos(
        np.sin(lat1)*np.sin(lat2) + 
        np.cos(lat1)*np.cos(lat2)*np.cos(lon2-lon1))

def airdistance(origin = "EBBR", destination = "EBBR"):
    if origin in airports and destination in airports:
        lat1 = airports[origin]["latitude"]
        lat2 = airports[destination]["latitude"]
        lon1 = airports[origin]["longitude"]
        lon2 = airports[destination]["longitude"]
        return int(distance(lat1, lon1, lat2, lon2))
    else:
        print("Not all airports known:", origin, "/",destination)
        return 0

def airdist(citypair):
    return airdistance(citypair["origin"], citypair["destination"])

There are some flights with incomplete origin info. The origin is passed as coordinates which we need to parse in a rudimentary way. 

In [81]:

def extract_coord(s):
    elem = s.split()
    t = (elem[1], elem[2]) # a tuple with two elements
    return t

def find_nearest(c):
    dist = 10000 # arbitrary high value
    airp = ""
    for a, aa in airports.items():
        d = distance(aa['latitude'], aa['longitude'], float(c[0]), float(c[1]))
        if d < dist:
            dist = d
            airp = a
    return airp

def validate_origin(o):
    if len(o)==4:    # rudimentary, but ICAO codes are 4 digits, good enough
        return o
    else:
        return find_nearest(extract_coord(o))

hist_flights['origin'] = hist_flights['origin'].apply(validate_origin)

def validate_city(city, icao):
    if not city:
        if icao in airports:
            city = airports[icao]['name'] # note that location is empty in the airports db, too
            print(city, icao)
    return city

hist_flights["originCity"] = hist_flights.apply(lambda r: validate_city(r['originCity'], 
                                                                       r['origin']), axis=1)
hist_flights["destinationCity"] = hist_flights.apply(lambda r: validate_city(r['destinationCity'], 
                                                                       r['destination']), axis=1)


Lome DXXX
Lome DXXX
Lome DXXX
N'Djili Int'l FZAA
Lome DXXX
Lome DXXX
Lome DXXX
Lome DXXX
Lome DXXX
Lome DXXX
Lome DXXX
Lome DXXX
Lome DXXX
Lome DXXX
Lome DXXX


As we will see, emergency flights were flown out of Liege, Antwerp, Frankfurt, and Zurich. In order to compare the emergency flights to the ordinary flights, we need to compare the "unique" destination, so precisely the other end of the citypair. For example, a flight DAKAR-FRANKFURT will count as a DAKAR flight.

In [82]:
origin = set(["EBBR", "EBLG", "EBAW", "EDDF", "LSZH"]) # special airports considered homebase
def decide_destination(flight):
    A = set([flight["origin"], flight["destination"]])
    B = set(["EBBR"]) # without the square brackets the set would contain E, B, R
    if A <= origin:   # is A wholly contained in list if 5?
        # then substract EBBR from it
        return A.difference(B).pop() # easy way to convert set of one element to element
    else: # at least one was not in the list, so remove the list to find it.
        # if both were in the list, select one arbitrarily
        return A.difference(origin).pop() # potentially one arbitrary of two elements
    
test_flight = {"origin": "LSZH", "destination" : "KIFR"}
decide_destination(test_flight)
    

'KIFR'

Apply the function defined above to both scheduled and historical flights.

In [83]:
hist_flights["udest"] = hist_flights.apply(decide_destination, axis=1)
bel_sched_flights["udest"] = bel_sched_flights.apply(decide_destination, axis=1)

In [85]:
hist_flights.loc[hist_flights['udest']=="GUCY"]

ADT             Filed  \
BEL1205-1459530459-airline-0098:0  2016-04-03 13:58  2016-04-02 15:38   

                                  actualarrivaltime actualdeparturetime  \
BEL1205-1459530459-airline-0098:0        1459711980          1459688317   

                                  aircrafttype destination destinationCity  \
BEL1205-1459530459-airline-0098:0         A333        GUCY         Conakry   

                                                 destinationName diverted  \
BEL1205-1459530459-airline-0098:0  Conakry Int'l (Gbessia Int'l)            

                                  estimatedarrivaltime  ...  filed_altitude  \
BEL1205-1459530459-airline-0098:0           1459711980  ...               0   

                                  filed_departuretime filed_ete  filed_time  \
BEL1205-1459530459-airline-0098:0          1459682100  06:00:00  1459607927   

                                     ident origin         originCity  \
BEL1205-1459530459-airline-0098:0  BEL1205   EDDF  Frankfurt am Main   

                                        originName route udest  
BEL1205-1459530459-airline-0098:0  Frankfurt Int'l        GUCY  

[1 rows x 23 columns]

In [86]:
citypairs["dist"] = 0
citypairs["dist"] = citypairs.apply(airdist, axis=1)


In [87]:
def available_seat_miles(row):
    origin = row['origin']
    destination = row['destination']
    seats = row['seats_cabin_coach']
    dist = airdistance(origin, destination)
    if not dist:
        print("WARNING: no dist for ", origin, " ", destination)
    return  dist * seats

bel_sched_flights["a_s_m"] = bel_sched_flights.apply(available_seat_miles, axis=1)


In [88]:
bel_sched_flights[bel_sched_flights['udest']=="GUCY"]

Empty DataFrame
Columns: [ident, aircrafttype, origin, destination, seats_cabin_coach, date, departuretime, udest, a_s_m]
Index: []

We will now define the relevant time period as 29 March to 5 April (included).

In [66]:
t1 = arrow.get(2016, 3, 29).timestamp
t2 = arrow.get(2016, 4, 6).timestamp

In [67]:
BSF = bel_sched_flights.loc[(bel_sched_flights['departuretime'] > t1) & 
                      (bel_sched_flights['departuretime'] < t2)]
np.floor(BSF.a_s_m.sum() / 1e6)

225.0

The total available seat miles for all city pairs was 225 mn.

### Top 20 Destinations by Available Seat Miles
We can now print the aggregate seat capacity (a_s_m) by city pair.

In [68]:
g1 = BSF.groupby(['udest'])

def myroundsum(x):
    return np.round(np.sum(x)/1e6, decimals=1)

sched_agg = g1.aggregate(myroundsum)
sched_agg = sched_agg.combine_first(ap).sort_values(by="a_s_m", ascending=False)
sched_agg.ix[0:19, [0, 3, 5]]

a_s_m                location                                       name
KJFK   14.9            New York, NY                        John F Kennedy Intl
FKKD   13.7                  Douala                               Douala Int'l
GOOY   12.4                    Yoff                                      Dakar
KIAD   11.1          Washington, DC                     Washington Dulles Intl
CYYZ   10.9        Toronto, Ontario                      Toronto Pearson Int'l
FZAA    9.8                Kinshasa                              N'Djili Int'l
LLBG    9.0                Tel Aviv                           Ben Gurion Int'l
DIAP    8.2                 Abidjan  Port Bouet (Felix Houphouet Boigny Int'l)
HRYR    6.5                  Kigali                               Kigali Int'l
DGAA    6.4                   Accra                               Kotoka Int'l
LEMD    5.8                  Madrid                              Barajas Int'l
HUEN    5.8  Entebbe (near Kampala)                              Entebbe Int'l
LIRF    5.1                    Rome  Leonardo da Vinci Int'l (Fiumicino Int'l)
LEMG    4.9                  Malaga                                     Malaga
GLRB    4.4                Monrovia                              Roberts Int'l
LEBL    4.4               Barcelona                            Barcelona Int'l
DFFD    4.1             Ouagadougou                                Ouagadougou
LFMN    4.0                    Nice                           Nice Cote d'Azur
FNLU    3.9                  Luanda                        Quatro de Fevereiro

Let's now take a look at the historical flights that were actually flown during the same time (29/3 to 5/4).

Merge the flight info with the aircraft info to obtain the number of seats per flight.
Subset the data to only include the date range we defined above.

In [69]:
hist_flights = hist_flights.join(aircraft, on="aircrafttype")

BHF = hist_flights.loc[(hist_flights['actualdeparturetime'] > t1) & 
                      (hist_flights['actualdeparturetime'] < t2), \
        ['ident','udest','ADT','origin','destination', 'originCity', 'destinationCity', 'seats_cabin_coach']]

In [70]:
def seat_miles_flown(row):
    origin = row['origin']
    destination = row['destination']
    seats = row['seats_cabin_coach']
    dist = airdistance(origin, destination)
    if not dist:
        print("WARNING: no dist for ", origin, " ", destination)
    return  dist * seats

BHF["smf"] = BHF.apply(seat_miles_flown, axis=1)
"Total seat miles flown in millions: {}".format(np.floor(BHF.smf.sum() / 1e6))

'Total seat miles flown in millions: 54.0'

There is one flight which appears twice in the data set (most likely an error):

In [71]:
bhf_g = BHF.groupby(["ident", "udest", "ADT"])
bhf_c = bhf_g.count()['smf']
bhf_c[bhf_c>1]

ident    udest  ADT             
BEL1205  GOOY   2016-04-03 21:50    2
Name: smf, dtype: int64

### Top 20 Destinations for Brussels Emergency Flights (by seat miles flown)

In [72]:

hist_agg = BHF.groupby(['udest']).aggregate(np.sum)
hist_agg.smf = hist_agg.smf.apply(lambda x: np.round(x / 1e6, decimals=1))
hist_agg.combine_first(ap).sort_values(by="smf", ascending=False).ix[0:19, [1, 3, 5]]

location                                       name  \
KJFK                New York, NY                        John F Kennedy Intl   
FZAA                    Kinshasa                              N'Djili Int'l   
FKYS                     Yaounde                     Yaounde Nsimalen Int'l   
HRYR                      Kigali                               Kigali Int'l   
GOOY                        Yoff                                      Dakar   
GFLL                    Freetown                                Lungi Int'l   
DIAP                     Abidjan  Port Bouet (Felix Houphouet Boigny Int'l)   
GLRB                    Monrovia                              Roberts Int'l   
DXXX                                                                   Lome   
LPPT                      Lisbon           Lisbon / Lisboa,Portela (Lisbon)   
FNLU                      Luanda                        Quatro de Fevereiro   
HUEN      Entebbe (near Kampala)                              Entebbe Int'l   
LFMN                        Nice                           Nice Cote d'Azur   
LEBL                   Barcelona                            Barcelona Int'l   
LPPR  Francisco Sa Carneiro Int.                             Porto / Oporto   
GUCY                     Conakry              Conakry Int'l (Gbessia Int'l)   
GCLP                Gran Canaria                         Gran Canaria Int'l   
DBBB                     Cotonou                        Cadjehoun (Cotonou)   
FKKD                      Douala                               Douala Int'l   

       smf  
KJFK  12.8  
FZAA   6.1  
FKYS   3.1  
HRYR   2.7  
GOOY   2.4  
GFLL   2.3  
DIAP   2.1  
GLRB   1.8  
DXXX   1.6  
LPPT   1.2  
FNLU   1.1  
HUEN   1.1  
LFMN   1.0  
LEBL   1.0  
LPPR   0.9  
GUCY   0.8  
GCLP   0.8  
DBBB   0.8  
FKKD   0.8

This routine shows how to obtain airport info from the flightaware API. Scheduled and historical flights were obtained in a similar fashion.

In [73]:
import requests
def AirportInfo(icao):
    username = 'REMOVED'
    apiKey = "REMOVED"
    baseURL = "http://flightxml.flightaware.com/json/FlightXML2/"
    method = "AirportInfo"
    data = {"airportCode" : icao}
    r = requests.post(baseURL+method, auth=(username, apiKey), data=data)
    if r.status_code == 200:
        res = r.json()
        return res['AirportInfoResult']
    else:
        print("Problem with airport:", icao)

AirportInfo("EGCC")

Problem with airport: EGCC


#### Comparison of Schedules (Regular versus Emergency)
We calculated the ratio of Seat Miles Flown / Available Seat Miles. The two top value reveals a data problem which we can explain by the fact that Lomé is normally only reached after a stop (so contributes little to ASM), but was directly flown to/from Frankfurt during the emergency schedule. Conakry, on the other hand, was not on the schedule, but for some reason was a stop during the emergeny schedule.

In [89]:
combo = sched_agg.combine_first(hist_agg)[['a_s_m', 'smf']]
combo["ratio"] = combo['smf']/combo['a_s_m']


It's interesting that Brussels kept flights to New York and achieved more than 85% of the scheduled capacity, but ditched Washington completely. More analysis would be required to explain this.

In [90]:
select = pd.notnull(combo["ratio"])
combo.combine_first(ap)[["name", "a_s_m", "smf", "ratio"]].sort_values(by="ratio", ascending=False)

name  a_s_m   smf     ratio
DXXX                                            Lome    0.2   1.6  8.000000
LIMF                    Turin Int'l (Torino Caselle)    0.1   0.1  1.000000
GBYD                     Banjul Int'l (Yundum Int'l)    0.1   0.1  1.000000
KJFK                             John F Kennedy Intl   14.9  12.8  0.859060
FZAA                                   N'Djili Int'l    9.8   6.1  0.622449
GFLL                                     Lungi Int'l    3.9   2.3  0.589744
GCLP                              Gran Canaria Int'l    1.6   0.8  0.500000
LGIR              Heraklion Int'l, Nikos Kazantzakis    0.4   0.2  0.500000
LIPE                     Bologna (Guglielmo Marconi)    0.6   0.3  0.500000
HRYR                                    Kigali Int'l    6.5   2.7  0.415385
GLRB                                   Roberts Int'l    4.4   1.8  0.409091
LGAV             Athens Int'l, Eleftherios Venizelos    1.6   0.6  0.375000
LPPT                Lisbon / Lisboa,Portela (Lisbon)    3.4   1.2  0.352941
LPPR                                  Porto / Oporto    2.7   0.9  0.333333
FNLU                             Quatro de Fevereiro    3.9   1.1  0.282051
DIAP       Port Bouet (Felix Houphouet Boigny Int'l)    8.2   2.1  0.256098
LFMN                                Nice Cote d'Azur    4.0   1.0  0.250000
LEBB                                          Bilbao    2.1   0.5  0.238095
LIML                                          Linate    1.7   0.4  0.235294
LFML                              Marseille Provence    1.3   0.3  0.230769
LEBL                                 Barcelona Int'l    4.4   1.0  0.227273
ENGM                                Oslo, Gardermoen    1.8   0.4  0.222222
LIRQ                     Amerigo Vespucci (Florence)    0.9   0.2  0.222222
LSGG                           Geneva Cointrin Int'l    3.4   0.7  0.205882
GOOY                                           Dakar   12.4   2.4  0.193548
HUEN                                   Entebbe Int'l    5.8   1.1  0.189655
DFFD                                     Ouagadougou    4.1   0.7  0.170732
EPKK                                    Krakow Int'l    0.6   0.1  0.166667
ESGG                                        Goteborg    1.9   0.3  0.157895
EDDT                                     Tegel Int'l    3.3   0.5  0.151515
...                                              ...    ...   ...       ...
EDDF                                 Frankfurt Int'l    NaN   0.1       NaN
EDDH                                         Hamburg    1.2   NaN       NaN
EGBB                                Birmingham Int'l    0.8   NaN       NaN
EGCC                                      Manchester    1.6   NaN       NaN
EGLL                                 London Heathrow    1.1   NaN       NaN
EKBI                                         Billund    0.2   NaN       NaN
EPWA                          Warsaw Frederic Chopin    1.6   NaN       NaN
FKYS                          Yaounde Nsimalen Int'l    NaN   3.1       NaN
GCTS                    Tenerife South (Reina Sofia)    1.1   NaN       NaN
GMAD                           Al Massira (Inezgane)    0.4   NaN       NaN
GMMX                                    Menara Int'l    0.7   NaN       NaN
GUCY                   Conakry Int'l (Gbessia Int'l)    NaN   0.8       NaN
HBBA                                 Bujumbura Int'l    1.0   NaN       NaN
KIAD                          Washington Dulles Intl   11.1   NaN       NaN
LDZA                                    Zagreb Pleso    0.8   NaN       NaN
LEAL                                        Alicante    0.9   NaN       NaN
LEAS                                        Asturias    NaN   NaN       NaN
LEMD                                   Barajas Int'l    5.8   NaN       NaN
LEZL                                       San Pablo    1.3   NaN       NaN
LFBD                               Bordeaux-Merignac    0.4   NaN       NaN
LFMM                                   Marseille FIR    NaN   0.2       NaN
LFPG                

#### Auxiliary Function to Investigate Flights to Certain Airports

In [99]:
def investigate(udest):
    print("--- SCHEDULED FLIGHTS ---")
    print(BSF.loc[BSF['udest'] == udest, ["date","ident", "origin", "destination"]])
    print("--- HISTORICAL FLIGHTS ---")
    print(BHF.sort_values(by="ADT").loc[BHF['udest'] == udest, ["ADT","ident", "origin", "destination"]])
    
investigate("GOOY")

--- SCHEDULED FLIGHTS ---
             date   ident origin destination
7043   2016-03-29  BEL206   GOOY        EBBR
7646   2016-03-29  BEL205   EBBR        GOOY
8210   2016-03-30  BEL203   EBBR        GOOY
8767   2016-03-30  BEL204   GOOY        EBBR
9327   2016-03-31  BEL205   EBBR        GOOY
9894   2016-03-31  BEL206   GOOY        EBBR
10043  2016-04-01  BEL204   GOOY        EBBR
11008  2016-04-01  BEL203   EBBR        GOOY
11213  2016-04-02  BEL203   EBBR        GOOY
11695  2016-04-02  BEL204   GOOY        EBBR
12066  2016-04-03  BEL205   EBBR        GOOY
12321  2016-04-03  BEL206   GOOY        EBBR
12902  2016-04-04  BEL203   EBBR        GOOY
13359  2016-04-04  BEL204   GOOY        EBBR
14013  2016-04-05  BEL205   EBBR        GOOY
14919  2016-04-05  BEL206   GOOY        EBBR
--- HISTORICAL FLIGHTS ---
                                                ADT    ident origin  \
BEL1204-1459241348-airline-0005:0  2016-03-31 00:35  BEL1204   GOOY   
BEL1203-1459515467-ed-0004:0       2016-

Let's use the k-means algorithm to categorize the airports into 3 continents, only using longitude and latitude. 

In [50]:
ap = pd.DataFrame(airports).T


In [109]:
import sklearn.cluster
k = sklearn.cluster.KMeans(n_clusters=3, init='k-means++', n_init=10, max_iter=300, tol=0.0001, 
                           precompute_distances='auto', verbose=0, random_state=None, copy_x=True, n_jobs=1)
df = ap[["latitude", "longitude"]]
df
ki = k.fit_predict(df)
ap["continent"] = ki

Continent 1 is Africa (but not Morocco)!

In [110]:
ap.loc[ap.continent==1]

latitude                location longitude  \
DBBB   6.35723                 Cotonou   2.38435   
DFFD   12.3532             Ouagadougou  -1.51242   
DGAA   5.60519                   Accra -0.166786   
DIAP   5.26139                 Abidjan  -3.92629   
DXXX   6.16561                           1.25451   
FKKD   4.00608                  Douala   9.71948   
FKYS   3.72256                 Yaounde   11.5533   
FNLU  -8.85838                  Luanda   13.2312   
FZAA  -4.38575                Kinshasa   15.4446   
GBYD    13.338                  Banjul  -16.6522   
GCLP   27.9319            Gran Canaria  -15.3866   
GCTS   28.0445                Tenerife  -16.5725   
GFLL   8.61644                Freetown  -13.1955   
GLRB   6.23379                Monrovia  -10.3623   
GOOY   14.7397                    Yoff  -17.4902   
GUCY   9.57689                 Conakry   -13.612   
HBBA  -3.32402               Bujumbura   29.3185   
HRYR  -1.96863                  Kigali   30.1395   
HUEN  0.042386  Entebbe (near Kampala)   32.4435   

                                           name             timezone  \
DBBB                        Cadjehoun (Cotonou)   :Africa/Porto-Novo   
DFFD                                Ouagadougou  :Africa/Ouagadougou   
DGAA                               Kotoka Int'l        :Africa/Accra   
DIAP  Port Bouet (Felix Houphouet Boigny Int'l)      :Africa/Abidjan   
DXXX                                       Lome         :Africa/Lome   
FKKD                               Douala Int'l       :Africa/Douala   
FKYS                     Yaounde Nsimalen Int'l       :Africa/Douala   
FNLU                        Quatro de Fevereiro       :Africa/Luanda   
FZAA                              N'Djili Int'l     :Africa/Kinshasa   
GBYD                Banjul Int'l (Yundum Int'l)       :Africa/Banjul   
GCLP                         Gran Canaria Int'l     :Atlantic/Canary   
GCTS               Tenerife South (Reina Sofia)     :Atlantic/Canary   
GFLL                                Lungi Int'l     :Africa/Freetown   
GLRB                              Roberts Int'l     :Africa/Monrovia   
GOOY                                      Dakar        :Africa/Dakar   
GUCY              Conakry Int'l (Gbessia Int'l)      :Africa/Conakry   
HBBA                            Bujumbura Int'l    :Africa/Bujumbura   
HRYR                               Kigali Int'l       :Africa/Kigali   
HUEN                              Entebbe Int'l      :Africa/Kampala   

      continent  
DBBB          1  
DFFD          1  
DGAA          1  
DIAP          1  
DXXX          1  
FKKD          1  
FKYS          1  
FNLU          1  
FZAA          1  
GBYD          1  
GCLP          1  
GCTS          1  
GFLL          1  
GLRB          1  
GOOY          1  
GUCY          1  
HBBA          1  
HRYR          1  
HUEN          1

Continent 0 is Europe, Israel and Morocco.

In [111]:
ap.loc[ap.continent==0]

latitude                            location longitude  \
EBAW  51.1894                    Antwerp / Deurne   4.46028   
EBBR  50.9014                            Brussels   4.48444   
EBLG  50.6374                               Liege   5.44322   
EDDF  50.0264                   Frankfurt am Main   8.54312   
EDDH  53.6304                             Hamburg   9.98823   
EDDT  52.5597                              Berlin   13.2877   
EGAC  54.6181           Belfast, Northern Ireland   -5.8725   
EGBB  52.4539                 Birmingham, England  -1.74803   
EGCC  53.3537                 Manchester, England  -2.27495   
EGLL  51.4775                     London, England -0.461389   
EGPH    55.95                 Edinburgh, Scotland   -3.3725   
EKBI  55.7403                             Billund   9.15178   
EKCH  55.6179                          Copenhagen    12.656   
ENGM  60.1939                                Oslo   11.1004   
EPKK  50.0777                              Balice   19.7848   
EPWA  52.1658                              Warsaw   20.9671   
ESGG  57.6628                          Landvetter   12.2798   
ESSB  59.3544                           Stockholm   17.9416   
EYVI  54.6341                             Vilnius   25.2858   
GMAD   30.325                              Agadir  -9.41307   
GMMX  31.6069                           Marrakech   -8.0363   
LDZA  45.7429                              Zagreb   16.0688   
LEAL  38.2822  Alicante / Benidorm / Costa Blanca -0.558156   
LEAS  43.5636  Aviles / Gijon / Oviedo (Asturias)  -6.03462   
LEBB  43.3011             Bilbao / Bilbo, Viscaya  -2.91061   
LEBL  41.2971                           Barcelona   2.07846   
LEMD  40.4936                              Madrid  -3.56676   
LEMG  36.6749                              Malaga  -4.49911   
LEZL   37.418                             Sevilla  -5.89311   
LFBD  44.8283                   Bordeaux/Merignac -0.715556   
LFBO  43.6291                             Blagnac   1.36382   
LFLL  45.7264                       Saint Exupery   5.09083   
LFML  43.4356                           Marseille   5.21361   
LFMM  43.6653                                 Aix     7.215   
LFMN  43.6607                                Nice   7.21758   
LFPG  49.0128                               Paris      2.55   
LFSB  47.5896                               Basel   7.52991   
LGAV  37.9364                              Athens   23.9445   
LGIR  35.3397         Heraklion (Iraklion), Crete   25.1803   
LHBP  47.4369                            Budapest   19.2556   
LICC  37.4668                             Catania   15.0664   
LICJ   38.176                                        13.091   
LIMC  45.6306                               Milan   8.72811   
LIMF  45.2008                      Turin (Torino)   7.64963   
LIML  45.4453                               Milan   9.27703   
LIPE  44.5349                             Bologna   11.2901   
LIPZ  45.5051                    Venice (Venezia)   12.3519   
LIRF  41.8045                                Rome   12.2508   
LIRN   40.886                              Naples   14.2908   
LIRQ    43.81                  Florence (Firenze)   11.2051   
LKPR  50.1008                              Prague     14.26   
LLBG  32.0095                            Tel Aviv   34.8769   
LOWW  48.1103                              Vienna   16.5697   
LPFR  37.0144            Faro / Algarve Int. Faro  -7.96591   
LPPR  41.2481          Francisco Sa Carneiro Int.  -8.68139   
LPPT  38.7813                              Lisbon  -9.13592   
LSGG  46.2381                              Geneva   6.10895   
LSZH  47.4647                              Zurich   8.54917   
ULLI  59.8003                    Saint Petersburg   30.2625   
UUDD  55.4088                              Moscow   37.9063   

                                                name            timezone  \
EBAW                                   Antwerp Int'l    :Europe/Brussels   
EBBR     

Determine which airports are closest to the cluster centers defined by k-means:

In [112]:
for kc in k.cluster_centers_:
    print(find_nearest(kc))

LIMC
DBBB
KIAD


The result was Milan, Cotonou, and Washington. Milan as a cluster center is interesting, and the map shows how Milan sits in the middle of the destinations in Europe. (The picture is from gcmap.com)
![Map](Milan-Center.png)

Can we use k-means clustering to find out something interesting about the emergency flights? So far, not. This will remain for another day...

In [113]:
import sklearn.cluster
co = combo.fillna(0)
k = sklearn.cluster.KMeans(n_clusters=3, init='k-means++', n_init=10, max_iter=300, tol=0.0001, 
                           precompute_distances='auto', verbose=0, random_state=None, copy_x=True, n_jobs=1)
ki = k.fit_predict(co)
ki

array([1, 2, 0, 0, 0, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 0, 2, 1,
       2, 0, 1, 2, 2, 0, 0, 2, 2, 1, 2, 2, 0, 0, 1, 1, 2, 2, 2, 0, 0, 0, 2,
       2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 0, 0, 2, 2, 2, 1,
       0, 2, 0, 0, 0, 2, 2, 2], dtype=int32)

In [114]:
combo["category"] = ki
combo

ask           smf     ratio  category
udest                                              
CYYZ   10907976.0           NaN       NaN         1
DBBB          NaN  7.762240e+05       NaN         2
DFFD    4510278.0  9.443720e+05  0.209382         0
DGAA    6632448.0  8.457520e+05  0.127517         0
DIAP    7814310.0  1.803254e+06  0.230763         0
DXXX          NaN  1.501576e+06       NaN         2
EDDF          NaN  5.216400e+04       NaN         2
EDDH    1247043.0           NaN       NaN         2
EDDT    3316420.0  4.916580e+05  0.148250         0
EGAC     414100.0  4.042600e+04  0.097624         2
EGBB     758336.0           NaN       NaN         2
EGCC    1649292.0           NaN       NaN         2
EGLL    1135734.0           NaN       NaN         2
EGPH    1486332.0  1.894200e+05  0.127441         2
EKBI     190404.0           NaN       NaN         2
EKCH    2987082.0  3.747400e+05  0.125454         0
ENGM    1766690.0  4.095179e+05  0.231800         2
EPKK     633672.0  1.111920e+05  0.175472         2
EPWA    1643936.0           NaN       NaN         2
ESGG    1899530.0  2.701080e+05  0.142197         2
ESSB    2983652.0  1.272640e+05  0.042654         0
EYVI     849888.0  1.027305e+05  0.120875         2
FKKD   13301472.0  7.485660e+05  0.056277         1
FKYS     422656.0  3.172648e+06  7.506454         2
FNLU    3918096.0  1.063244e+06  0.271368         0
FZAA    9800976.0  6.052102e+06  0.617500         1
GCLP    1578168.0  7.948800e+05  0.503673         2
GCTS    1067568.0           NaN       NaN         2
GFLL    4272280.0  2.511836e+06  0.587938         0
GLRB    4090670.0  1.603756e+06  0.392052         0
...           ...           ...       ...       ...
LFBO    1811544.0  8.938000e+04  0.049339         2
LFLL    1828080.0  9.348000e+04  0.051136         2
LFML    1276740.0  2.656800e+05  0.208092         2
LFMM          NaN  1.897972e+05       NaN         2
LFMN    3982806.0  1.024356e+06  0.257195         0
LFPG     346656.0           NaN       NaN         2
LFSB     483472.0           NaN       NaN         2
LGAV    1591356.0  6.039805e+05  0.379538         2
LGIR     413172.0  2.003760e+05  0.484970         2
LHBP    3869262.0           NaN       NaN         0
LICC    1045482.0           NaN       NaN         2
LICJ     229320.0           NaN       NaN         2
LIMC    1474080.0  2.189229e+05  0.148515         2
LIMF      69208.0  6.920800e+04  1.000000         2
LIML    1727026.0  4.392585e+05  0.254344         2
LIPE     617984.0  2.755200e+05  0.445837         2
LIPZ    2831328.0  2.644080e+05  0.093387         0
LIRF    5072760.0  1.624031e+05  0.032015         0
LIRN    1375920.0  1.872082e+05  0.136060         2
LIRQ     863460.0  1.974560e+05  0.228680         2
LKPR    2251560.0  3.095900e+05  0.137500         2
LLBG    8968896.0  5.478600e+05  0.061084         1
LOWW    3330436.0  4.915949e+05  0.147607         0
LPFR    2431980.0           NaN       NaN         2
LPPR    2699280.0  9.005708e+05  0.333634         0
LPPT    3418578.0  1.203360e+06  0.352006         0
LSGG    3413292.0  6.582389e+05  0.192846         0
LSZH          NaN  2.008171e+05       NaN         2
ULLI     828360.0           NaN       NaN         2
UUDD    2101056.0           NaN       NaN         2

[77 rows x 4 columns]